### ISSUE TITLE
@tus/s3-store: add object prefixing

### ISSUE THREAD

**Issue by Murderlon** [tus/tus-node-server#457]
Prefixing can be used to create a pseudo-directory structure in the bucket.

tusd supports this with a hardcoded `ObjectPrefix` option:

https://github.com/tus/tusd/blob/b4ffdf43f034099ee46437271de2ff98c1f52397/pkg/s3store/s3store.go#L987-L994

in @tus/s3-store we have `partKey` where this could be supported:

https://github.com/tus/tus-node-server/blob/0af377dfc4877b44c95238a3b46d96429cad96ca/packages/s3-store/index.ts#L144-L154

However it might be more flexible have a function `getObjectPrefix (upload: Upload) => string` so it can be done dynamically based on metadata?

cc @fenos

**Comment by Murderlon** [comment]
I'm not sure that would be better. We already have [`namingFunction`](https://github.com/tus/tus-node-server/tree/main/packages/server#optionsnamingfunction) which would more or less be conflicting. It also feels a bit hacky to use the ID to create a storage path implicitly, rather than explicitly prepending it with a object prefix for S3, and the fact that we would have to reverse whatever was done to generate the URL.

If it's valuable to have the `req` for this function, we can make that happen too in `getObjectPrefix` on the store itself.

cc @Acconut

**Comment by fenos** [comment]
Yea, I agree that doing it this way seems a bit hacky, somewhat seems like a workaround on the TUS protocol to support implicit paths.

I believe that the use case is quite legitimate to have implicit paths, and it can apply to any sort of Storage adapter.

In our case, we also support the FileStore adapter which we would need the same object prefixing functionality in there, hence I thought that having it at the server level could make sense since you want this prefix shared across all store adapters.

I agree that `namingFunction` and `getObjectPrefix` would be quite conflicting, maybe we can simply expose the two methods `getFileIdFromRequest` and `generateUrl` on the Server instead and give full freedom to the user:

```ts
const server = new Server({
   generateUrl(): string {
      ....
   },
   getFileIdFromRequest(): string {
      ...
   }
})
```

This way the user can have the freedom of customising the URL to its liking and still be consistent to the underline TUS protocol

**Comment by Acconut** [comment]
I agree that it would be useful to be able to separate the upload ID (and thus the upload URL) from the storage path. Sometimes you want to customize the upload URL, sometimes the storage path, and sometimes both. It's not the most pressing issue, so the first approach for implementation was to just couple the two together.

We could thing about a general solution where the user is able to fully influence the storage path for all storages, but I think adding an option for object prefix is also fine. That's the easier option for users if they don't want to fiddle around with a custom storage path generation approach.

> TUS protocol which maps URI 1:1 to the file path

It's worth noting that the tus protocol itself does not make any assumption about the URL format or the storage path. They can be whatever they want. Only in our implementation (tusd and tus-node-server) we have added the concept of an upload ID that corresponds to the destination because it is a simple implementation that works great in most situations.

**Comment by Murderlon** [comment]
I like the idea of splitting it into two functions. However it puts quite the burden on on the implementer and would make `options.relativeLocation` and `options.respectForwardedHeaders` obsolete  when used. People would have to reimplement this logic:

https://github.com/tus/tus-node-server/blob/0af377dfc4877b44c95238a3b46d96429cad96ca/packages/server/src/handlers/BaseHandler.ts#L35-L70

Those are complexities I'd rather keep internal. If the problem is specifically about augmenting _storage paths_, then think we should keep solution tied to the storage solutions rather than make the ID's "magical".

We can do something like this:

```js
const {Server} = require('@tus/server')
const {FileStore} = require('@tus/file-store')
const {S3Store} = require('@tus/s3-store')

const getFilePath = (req, upload) => req.headers['tenant-id']
const s3Store = new S3Store({getFilePath})
const fileStore = new FileStore({getFilePath})
const datastore = useFileStore ? fileStore : s3Store

const server = new Server({path: '/files', datastore})
```

**Comment by Acconut** [comment]
Alternatively, you could also generate the storage path once before upload is created and then store it in the upload's info object persistently. Whenever, you load an upload, you first look up the info file, extract the storage path and then load the upload from there.

The benefit is that you do not have to derive the storage path from the upload ID for every request. The downside is that this way the info file cannot have a custom path.

**Comment by shellscape** [comment]
To toss another variant in here: We'd like to store the metadata in a different location/prefix than the resulting file.

**Comment by elliotdickison** [comment]
@Murderlon Would you mind giving an example of how to use these functions to set an object prefix in S3? With @tus/server@1.0.0 and @tus/s3-store@1.0.1 I was able to set an object prefix using the `namingFunction` option and it worked perfectly. After upgrading to 1.2.0 for both packages I am now getting "No file found" 404 errors for the PATCH request that follows upload creation.

I tried moving my prefix logic from `namingFunction` to `getFileIdFromRequest` but I still get a 404 for the PATCH request. In fact I get a 404 for the patch request if I use `getFileIdFromRequest` at all.  I moved my prefix logic to `generateUrl` to customize the URL which worked as far as changing the URL goes, but the file is uploaded to S3 without the prefix present in that URL.

Any ideas?

**Comment by Murderlon** [comment]
Hi, thanks for the feedback. There definitely should not have been a regression so I'll take a look. I'll see what I can do about the docs as well to clear it up and perhaps we want to think about deprecating `namingFunction` to only have two functions.

**Comment by Murderlon** [comment]
@elliotdickison do you have a reproducible example? I created a test and ran it against the latest version and on `@tus/server@1.0.0` and they both fail when `namingFunction` returns an extra path prefix. Both versions succeed if you return something without a `/`. 

```ts
    it.only('should use namingFunction correctly', (done) => {
      const server = new Server({
        path: '/test/output',
        datastore: new FileStore({directory: './test/output'}),
        namingFunction() {
          return '1234/5678'
        },
      })
      const length = Buffer.byteLength('test', 'utf8').toString()
      request(server.listen())
        .post(server.options.path)
        .set('Tus-Resumable', TUS_RESUMABLE)
        .set('Upload-Length', length)
        .then((res) => {
          console.log(res.headers.location)
          request(server.listen())
            .patch(removeProtocol(res.headers.location))
            .send('test')
            .set('Tus-Resumable', TUS_RESUMABLE)
            .set('Upload-Offset', '0')
            .set('Content-Type', 'application/offset+octet-stream')
            .expect(204, done)
        })
    })

```

**Comment by elliotdickison** [comment]
Thanks for the quick response and updated docs, those did help. Now that I understand the functions I've pinpointed the problem I'm having: `namingFunction` can no longer return the extra path prefix, and `generateUrl` only affects the `Location` header but has no bearing on the object's key.

I used to be able to do this:
```
// Used to work, no longer
new Server({
  namingFunction () {
    return "file/will/be/stored/here.txt"
  }
})
```

Now if I do this the object key is equivalent to `namingFunction` and `generateUrl` seems to have no practical effect:
```
new Server({
  // The object key will be here.txt
  namingFunction () {
    return "here.txt"
  },
  // The Location header will be /file/will/be/stored/here.txt but this does not affect
  // the final object key in storage. The extra prefix "/file/will/be/stored" will be ignored
  // and the full object key will be "here.txt"
  generateUrl(req, { baseUrl, id }) {
    return `${baseUrl}/file/will/be/stored/${id}`
  }
})
```

**Comment by elliotdickison** [comment]
Now that I'm reading the description of this PR again though ("Prefixing can be used to create a pseudo-directory structure in the bucket."), what is the PR about if not allowing slashes in the object key? Am I misunderstanding the discussion here?

**Comment by elliotdickison** [comment]
@fenos That did the trick! Thank you very much.

**Comment by Murderlon** [comment]
I'll take a look next week at supporting this in file store too and updating the docs to include the above example.

**Comment by Murderlon** [comment]
Added in this PR now: https://github.com/tus/tus-node-server/pull/549

**Comment by mkabatek** [comment]
> This is important to note, read [example-store-files-in-custom-nested-directories](https://www.npmjs.com/package/@tus/server#example-store-files-in-custom-nested-directories) and setup accordingly. I leave this here for others to find.

Thank you all and @lalilaloe thank you so much for adding this nugget of information to get custom file paths! So much appreciated!

**Comment by Acconut** [comment]
> Would be amazing if `tusd` supported this out of the box.

Thanks to https://github.com/tus/tusd/pull/1083, the latest release candidate of tusd support fully customized paths for the file store. Soon (see https://github.com/tus/tusd/pull/1167) the same will be possible for the S3 store, followed by Azure and Google Cloud Storage support :)

**Comment by HG2407** [comment]
@Acconut any updates on supporting this for gcs ?

**Comment by brad07** [comment]
@fenos  @lalilaloe Thanks so much for solving this and documenting it. It was a life saver for me.
